cpp/src/arrow/compute/api_vector.h

```diff
@@ -705,5 +736,56 @@ Result<std::shared_ptr<Array>> PairwiseDiff(const Array& array,
                                             bool check_overflow = false,
                                             ExecContext* ctx = NULLPTR);
 
+/// \brief Return the reverse indices of the given indices.
+///
+/// For indices[i] = x, reverse_indices[x] = i. And reverse_indices[x] = null if x does
+/// not appear in the input indices. For indices[i] = x where x < 0 or x >= output_length,
+/// it is ignored. If multiple indices point to the same value, the last one is used.
+///
+/// For example, with indices = [null, 0, 3, 2, 4, 1, 1], the reverse indices is
+///   [1, 6, 3]                    if output_length = 3,
+///   [1, 6, 3, 2, 4, null, null]  if output_length = 7.
```

Comment:
It is particularly useful for cases where the indices are only a subset of the output (in other words, the output contains "holes"/`null`s). Consider an input `[3, 0]` (imagine that row `1` and `2` are evaluated to `null` in the `if_else` special form's condition so they go to neither `true` nor `false` branch, thus having no slots in the input to `reverse_indices`/`permute`), it is desired to have the output as `[1, null, null, 0]` (as opposed to a presumable length-2 output `[1, null]`).

This is also noted in my comments in one of the tests: https://github.com/apache/arrow/pull/44394/files#diff-ee27abfb87a9105ebbdb8bdcdd26aaec752f2631181a078e3ef5a07b4f00d1ceR995-R1000

Reply:
Right, so could you call it `n` or `max_index` (i.e. `n - 1`) instead. The output length becomes a consequence of how far the loop goes. So at least to me it was easier to understand this parameter as a constraint on the loop and the output length being just a consequence of that.

Reply:
Another name suggestion: `stop_index` instead of `n`. That would also reduce the weirdness around passing a negative `output_length` (lenghts can't be negative), but a negative `stop_index` makes sense.

Reply:
I'll take `max_index` then. I think `stop_index` gives people an impression that the processing "stops early" if no index in the input reaches it, which is against the fact that in such case tail nulls will be appended. Whereas `max_index` can imply such tail nulls slightly.

Thank you for the elaboration.